rust/debian/Dockerfile

```diff
@@ -34,24 +34,175 @@ RUN --mount=type=secret,id=aws,target=/root/.aws/credentials \
+    echo "Building zlib" && \
+    cd /tmp/zlib/zlib-$ZLIB_VERSION && \
+    export CC="$CC_WRAPPER clang --target=${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl" \
+    CFLAGS="-nostdinc -isystem $LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl/include/" \
+    LDSHARED="lld" AR=llvm-ar RANLIB=llvm-ranlib \
+    LDFLAGS="-L$LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl/lib"; \
+    # Legacy path \
+    ./configure --static --shared --prefix=/usr/local/musl || cat configure.log && \
+    make install && \
+    # sysroot \
+    ./configure --static --shared --prefix=$LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl || cat configure.log && \
+    make install && \
+    sccache --show-stats && \
+    rm -r /tmp/zlib
+
+
+## Install static -fPIC Rocksdb
+RUN --mount=type=secret,id=aws,target=/root/.aws/credentials \
+    if [ -z "$SCCACHE_BUCKET" ]; then unset CC_WRAPPER; fi; \
+    if [ -z "$SCCACHE_ENDPOINT" ]; then unset SCCACHE_ENDPOINT; fi; \
+    echo "Building rocksdb" && \
+    git clone https://github.com/facebook/rocksdb.git --branch v6.22.1 --depth=1 /tmp/rocksdb; \
+    cd /tmp/rocksdb && \
+    export MAKEFLAGS="--silent -j $(nproc)"; \
+    export USE_CLANG=1 TARGET_OS=Linux PORTABLE=1 AR=llvm-ar; \
+    export TARGET_ARCHITECTURE=${CROSS_COMPILER_TARGET_ARCH}; \
+    export C_INCLUDES="-isystem $LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl/include/ -isystem $LLVM_SYSROOT/usr/lib/clang/${LLVM_VERSION}/include/ -isystem $LLVM_SYSROOT/usr/include/"; \
+    export CC="$CC_WRAPPER clang --target=${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl"; \
+    export EXTRA_CFLAGS="-nostdinc -fPIC $C_INCLUDES -nostdlib -nostdlib++ -nodefaultlibs -Wno-unused-command-line-argument"; \
+    export CXX="$CC_WRAPPER clang++ -nostdlib++ -fuse-ld=lld --target=${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl"; \
+    export EXTRA_CXXFLAGS="-nostdinc -nostdinc++ -fPIC -isystem $LLVM_SYSROOT/usr/include/c++/v1/ -isystem $LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl/include/ -D _LIBCPP_HAS_MUSL_LIBC $C_INCLUDES -Wno-unused-command-line-argument "; \
+    export CFLAGS="$EXTRA_CFLAGS -L$LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl/lib"; \
+    DISABLE_WARNING_AS_ERROR=1 make static_lib && \
+    cp librocksdb.a* $LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl/lib && \
+    cp -r include/* $LLVM_SYSROOT/usr/${CROSS_COMPILER_TARGET_ARCH}-unknown-linux-musl/include/ && \
+    sccache --show-stats && \
+    rm -R /tmp/rocksdb/
+
+
 #############################################################################
 ## The Cross Compile image
 #############################################################################
 FROM --platform=${TARGETPLATFORM} ${BASE_IMAGE}
 
 # Copy in the cross compiled cargo tools from base
 COPY --from=build /out/tools/* /usr/local/bin/
```

Comment:
re: dropins

how do we control compatibility between build and runtime (image) environments?  to match api, headers, api vers etc.

ideal case:  build where it runs, or alternatively use exactly the same env during build as during runtime.

Reply:
The cargo tools are all compiled against musl, so there aren't any runtime dependencies except the linux syscall abi.

The libc++, rocksdb.a and libc are all configured and built against the musl libc too and the compiler-builtins libs provide all of the isa specific implementations of primitives. Those are all used for the linux static binaries, none of this is used by the k8s image binaries